lib/ddtrace/contrib/configuration/integrations.rb

```diff
@@ -0,0 +1,106 @@
+require 'ddtrace/contrib/configuration/integration_definition_set'
+
+module Datadog
+  module Contrib
+    module Configuration
+      # Behavior for a configuration object that contains integration settings
+      module Integrations
+        def self.included(base)
+          base.send(:extend, ClassMethods)
+          base.send(:include, InstanceMethods)
+        end
+
+        # Class behavior for a configuration object that contains integration settings
+        module ClassMethods
+          def integrations
+            @integrations ||= begin
+              # Allows for class inheritance of integration definitions
+              superclass <= Integrations ? superclass.integrations.dup : IntegrationDefinitionSet.new
+            end
+          end
+
+          protected
+
+          def integration(name)
+            assert_valid_integration!(name)
+            integrations[name] = IntegrationDefinition.new(name).tap do
+              define_integration_accessors(name)
+            end
+          end
+
+          private
+
+          def assert_valid_integration!(name)
+            unless Datadog.registry[name]
+              raise(InvalidIntegrationDefinitionError, "Registry doesn't define the integration: #{name}")
+            end
+          end
+
+          def define_integration_accessors(name)
+            integration_name = name
+
+            define_method(integration_name) do |*args, &block|
+              configure_integration(integration_name, *args, &block)
+            end
+          end
+        end
+
+        # Instance behavior for a configuration object with options
+        module InstanceMethods
+          # Applies configuration and activates an integration
+          # TODO: Implement #use/#set/#[] from Extensions instead
+          def activate!(name, *args, &block)
+            integration = get_integration(name)
+            integration.activate!(*args, &block)
+          end
+
+          def integrations
+            @integrations ||= IntegrationSet.new
+          end
+
+          def configure_integration(name, *args, &block)
+            integration = get_integration(name)
+            integration.configure(*args, &block)
+          end
+
+          def get_integration(name)
+            add_integration(name) unless integrations.key?(name)
+            integrations[name]
+          end
+
+          def integrations_hash
+            integrations.each_with_object({}) do |(key, _), hash|
+              hash[key] = get_integration(key)
+            end
+          end
+
+          def reset_integrations!
+            integrations.values.each(&:reset)
+          end
+
+          private
+
+          def add_integration(name)
+            assert_valid_integration!(name)
+            definition = self.class.integrations[name]
+            Integration.new(definition).tap do |integration|
+              integrations[name] = integration
+            end
+          end
+
+          def assert_valid_integration!(name)
```

Comment:
Here the collision between our two definitions of "integration" seems rather confusing:
We have two `assert_valid_integration!`, one instance method and one class method: the instance method refers to our new "component" integrations, while the class method refers to our top-level integrations.
This seems to reinforce the point that we probably shouldn't keep both concepts using the same name.

Reply:
I don't think these are actually collisions; the class version is verifying that the definition is correct e.g. `integration :foo` is actually an integration at all, and the instance version verifies whether this particular integration is defined as a component of the parent e.g. `c.use :rails { |rails| rails.activate!(:sinatra) }` should not allow a user to activate Sinatra as a component of Rails.

Reply:
I was somewhat confused while following the code for the first time, but I agree that both implementation and naming are correct.
I'm ok leaving them as is.